<p><a href="https://github.com/PMBio/MOFA">MOFA</a> was used to analyze RNA-seq data and ChIP-seq data for 3 histone marks together, looking for sources of variation shared acorss all data types. Two different analyses were done, one with ChIP-seq reads counted in promoter regions, and one with ChIP-seq reads counted in called peaks. Each analysis is split into two steps. The first step (“run”) is fitting the model (which takes a long time) and performing some basic QC, while the second step (“analyze”) loads the fitted model and performs some analysis.</p>
<h2 id="peak-based-analysis">Peak-based analysis</h2>
<ul>
<li>Step 1: <a href="peak-mofa-run.html"><code>peak-mofa-run.html</code></a></li>
<li>Step 2: <a href="peak-mofa-analyze.html"><code>peak-mofa-analyze.html</code></a></li>
</ul>
<h2 id="promoter-based-analysis">Promoter-based analysis</h2>
<ul>
<li>Step 1: <a href="promoter-mofa-run.html"><code>promoter-mofa-run.html</code></a></li>
<li>Step 2: <a href="promoter-mofa-analyze.html"><code>promoter-mofa-analyze.html</code></a></li>
</ul>
